src/storage/restrict.rs

```diff
@@ -0,0 +1,224 @@
+
+use std::marker::PhantomData;
+use std::ops::DerefMut;
+use std::borrow::{Borrow, BorrowMut};
+use std::fmt;
+
+use hibitset::BitSet;
+
+use storage::MaskedStorage;
+use world::EntityIndex;
+use {Component, DistinctStorage, Entities, Entity, Index, Join, ParJoin, Storage, UnprotectedStorage};
+
+/// Specifies that the `RestrictedStorage` cannot run in parallel.
+pub enum NormalRestriction { }
+/// Specifies that the `RestrictedStorage` can run in parallel.
+pub enum ParallelRestriction { }
+
+/// Similar to a `MaskedStorage` and a `Storage` combined, but restricts usage
+/// to only getting and modifying the components. That means nothing that would
+/// modify the inner bitset so the iteration cannot be invalidated. For example,
+/// no insertion or removal is allowed.
+pub struct RestrictedStorage<'rf, 'st: 'rf, B, T, R, RT>
+    where T: Component,
+          R: Borrow<T::Storage> + 'rf,
+          B: Borrow<BitSet> + 'rf,
+{
+    bitset: B,
+    data: R,
+    entities: &'rf Entities<'st>,
+    phantom: PhantomData<(T, RT)>,
+}
+
+unsafe impl<'rf, 'st: 'rf, B, T, R> ParJoin for &'rf mut RestrictedStorage<'rf, 'st, B, T, R, ParallelRestriction>
+    where T: Component,
+          R: BorrowMut<T::Storage> + 'rf,
+          B: Borrow<BitSet> + 'rf,
+{ }
+
+impl<'rf, 'st, B, T, R, RT> RestrictedStorage<'rf, 'st, B, T, R, RT>
+    where T: Component,
+          R: Borrow<T::Storage>,
+          B: Borrow<BitSet>,
+{
+    /// Attempts to get the component related to the entity.
+    ///
+    /// Functions similar to the normal `Storage::get` implementation.
+    pub fn get(&self, entity: Entity) -> Option<&T> {
+        if self.bitset.borrow().contains(entity.id()) && self.entities.is_alive(entity) {
+            Some(unsafe { self.data.borrow().get(entity.id()) })
+        } else {
+            None
+        }
+    }
+
+    /// Gets the component related to the current entry without checking whether
+    /// the storage has it or not.
+    pub fn get_unchecked(&self, entry: &Entry<'rf, T>) -> &T {
+        entry.assert_same_storage(self.data.borrow());
+        unsafe { self.data.borrow().get(entry.index()) }
+    }
+}
+
+impl<'rf, 'st, B, T, R, RT> RestrictedStorage<'rf, 'st, B, T, R, RT>
+    where T: Component,
+          R: BorrowMut<T::Storage>,
+          B: Borrow<BitSet>,
+{
+    /// Gets the component related to the current entry without checking whether
+    /// the storage has it or not.
+    pub fn get_mut_unchecked(&mut self, entry: &Entry<'rf, T>) -> &mut T {
```

Comment:
This is actually safe since it requires an `Entry`, which can't live longer than the iteration.